### ISSUE TITLE
Syncgateway is not getting started. panic: runtime error: assignment to entry in nil map

### ISSUE THREAD

**Issue by keshavkatwe** [couchbase/sync_gateway#1015]
11:43:43.742135 Starting admin server on 127.0.0.1:4985
11:43:43.749574 Starting server on :4984 ...
panic: runtime error: assignment to entry in nil map

goroutine 50 [running]:
runtime.panic(0x903120, 0x11cb93d)
    /usr/local/go/src/pkg/runtime/panic.c:266 +0xb6
github.com/couchbaselabs/sync_gateway/db.func·019(0xc21021d600, 0x58, 0xc210266004, 0x257)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/shadower.go:121 +0x3b0
github.com/couchbaselabs/sync_gateway/db.func·013(0xc210266004, 0x257, 0x257, 0x0, 0x0, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/crud.go:472 +0x283
github.com/couchbaselabs/sync_gateway/base.func·001(0xc210266004, 0x257, 0x257, 0x0, 0x0, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/base/bucket.go:67 +0x7a
github.com/couchbaselabs/go-couchbase.func·013(0xc210266004, 0x257, 0x257, 0x0, 0x0, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/client.go:584 +0x92
github.com/couchbaselabs/go-couchbase.(_Bucket).update(0xc2100ac780, 0xc21024fc00, 0x58, 0x0, 0x7ffaca7a6aa8, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/client.go:565 +0xac
github.com/couchbaselabs/go-couchbase.(_Bucket).WriteUpdate(0xc2100ac780, 0xc21024fc00, 0x58, 0x0, 0x7ffaca7a6b08, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/client.go:589 +0xa6
github.com/couchbaselabs/sync_gateway/base.couchbaseBucket.WriteUpdate(0xc2100ac780, 0xc21024fc00, 0x58, 0x0, 0xc21029f700, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/base/bucket.go:71 +0x7a
github.com/couchbaselabs/sync_gateway/db.(_Database).updateDoc(0xc2101c2280, 0xc21024fc00, 0x58, 0x928200, 0xc2102a3e40, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/crud.go:596 +0x369
github.com/couchbaselabs/sync_gateway/db.(_Shadower).pullDocument(0xc2101a1a40, 0xc21024fc00, 0x58, 0xc2102b39f1, 0x4, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/shadower.go:132 +0x20a
github.com/couchbaselabs/sync_gateway/db.(*Shadower).readTapFeed(0xc2101a1a40)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/shadower.go:77 +0x275
created by github.com/couchbaselabs/sync_gateway/db.NewShadower
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/shadower.go:36 +0x117

goroutine 1 [IO wait]:
net.runtime_pollWait(0x7ffaca94db78, 0x72, 0x0)
    /usr/local/go/src/pkg/runtime/netpoll.goc:116 +0x6a
net.(_pollDesc).Wait(0xc2101bbed0, 0x72, 0x7ffaca94c0c0, 0xb)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:81 +0x34
net.(_pollDesc).WaitRead(0xc2101bbed0, 0xb, 0x7ffaca94c0c0)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:86 +0x30
net.(_netFD).accept(0xc2101bbe70, 0xab9170, 0x0, 0x7ffaca94c0c0, 0xb)
    /usr/local/go/src/pkg/net/fd_unix.go:382 +0x2c2
net.(_TCPListener).AcceptTCP(0xc2101bea78, 0x18, 0xc210043050, 0x473f43)
    /usr/local/go/src/pkg/net/tcpsock_posix.go:233 +0x47
net.(_TCPListener).Accept(0xc2101bea78, 0x50, 0x9b6800, 0x3, 0x998900)
    /usr/local/go/src/pkg/net/tcpsock_posix.go:243 +0x27
net/http.(_Server).Serve(0xc2101bc0a0, 0x7ffaca9573d8, 0xc2101bea78, 0x0, 0x0)
    /usr/local/go/src/pkg/net/http/server.go:1622 +0x91
github.com/couchbaselabs/sync_gateway/base.ListenAndServeHTTP(0x998900, 0x5, 0x0, 0x0, 0x0, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/base/http_listener.go:52 +0x3ee
github.com/couchbaselabs/sync_gateway/rest.(*ServerConfig).serve(0xc210083300, 0x998900, 0x5, 0x7ffaca94c2e0, 0xc2101d32c0)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/rest/config.go:390 +0xa7
github.com/couchbaselabs/sync_gateway/rest.RunServer(0xc210083300)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/rest/config.go:430 +0x6e9
github.com/couchbaselabs/sync_gateway/rest.ServerMain()
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/rest/config.go:436 +0x2b
main.main()
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/main.go:16 +0x1a

goroutine 23 [sleep]:
time.Sleep(0x9502f900)
    /usr/local/go/src/pkg/runtime/time.goc:31 +0x31
github.com/couchbaselabs/sync_gateway/db.func·003()
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/change_cache.go:68 +0x62
created by github.com/couchbaselabs/sync_gateway/db.(*changeCache).Init
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/change_cache.go:70 +0x1be

goroutine 24 [select]:
github.com/couchbaselabs/go-couchbase.(_TapFeed).run(0xc21007eec0)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/tap.go:55 +0x2f9
created by github.com/couchbaselabs/go-couchbase.(_Bucket).StartTapFeed
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/tap.go:39 +0x14a

goroutine 25 [chan receive]:
github.com/couchbaselabs/sync_gateway/base.func·002()
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/base/bucket.go:94 +0x53
created by github.com/couchbaselabs/sync_gateway/base.couchbaseBucket.StartTapFeed
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/base/bucket.go:104 +0x1bd

goroutine 26 [chan receive]:
github.com/couchbaselabs/sync_gateway/db.func·006()
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/change_listener.go:49 +0xb5
created by github.com/couchbaselabs/sync_gateway/db.(*changeListener).Start
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/change_listener.go:66 +0x25e

goroutine 27 [chan receive]:
github.com/couchbaselabs/sync_gateway/db.(*DatabaseContext).watchDocChanges(0xc2100b2b40)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/assimilator.go:16 +0xa5
created by github.com/couchbaselabs/sync_gateway/db.NewDatabaseContext
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/db/database.go:114 +0x339

goroutine 29 [IO wait]:
net.runtime_pollWait(0x7ffaca94dc20, 0x72, 0x0)
    /usr/local/go/src/pkg/runtime/netpoll.goc:116 +0x6a
net.(_pollDesc).Wait(0xc2100d08b0, 0x72, 0x7ffaca94c0c0, 0xb)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:81 +0x34
net.(_pollDesc).WaitRead(0xc2100d08b0, 0xb, 0x7ffaca94c0c0)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:86 +0x30
net.(_netFD).Read(0xc2100d0850, 0xc21016eda0, 0x18, 0x18, 0x0, ...)
    /usr/local/go/src/pkg/net/fd_unix.go:204 +0x2a0
net.(_conn).Read(0xc21016b298, 0xc21016eda0, 0x18, 0x18, 0x25, ...)
    /usr/local/go/src/pkg/net/net.go:122 +0xc5
io.ReadAtLeast(0x7ffaca94dcf0, 0xc21016b298, 0xc21016eda0, 0x18, 0x18, ...)
    /usr/local/go/src/pkg/io/io.go:288 +0xf6
io.ReadFull(0x7ffaca94dcf0, 0xc21016b298, 0xc21016eda0, 0x18, 0x18, ...)
    /usr/local/go/src/pkg/io/io.go:306 +0x71
github.com/dustin/gomemcached.(_MCRequest).Receive(0xc21017d900, 0x7ffaca94dcf0, 0xc21016b298, 0xc21016eda0, 0x18, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/dustin/gomemcached/mc_req.go:129 +0xdf
github.com/dustin/gomemcached/client.(_Client).runFeed(0xc210082810, 0xc21017d300, 0xc21016ed60)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/dustin/gomemcached/client/tap_feed.go:270 +0x11c
created by github.com/dustin/gomemcached/client.(*Client).StartTapFeed
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/dustin/gomemcached/client/tap_feed.go:251 +0x215

goroutine 30 [select]:
github.com/couchbaselabs/go-couchbase.(_TapFeed).forwardTapEvents(0xc21007eec0, 0xc21016ed60, 0xc21005f240, 0xc2100fc8c0, 0x11)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/tap.go:101 +0x249
created by github.com/couchbaselabs/go-couchbase.(_TapFeed).connectToNodes
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/tap.go:93 +0x367

goroutine 48 [select]:
github.com/couchbaselabs/go-couchbase.(_TapFeed).run(0xc2101a1a00)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/tap.go:55 +0x2f9
created by github.com/couchbaselabs/go-couchbase.(_Bucket).StartTapFeed
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/tap.go:39 +0x14a

goroutine 49 [runnable]:
github.com/couchbaselabs/sync_gateway/base.func·002()
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/base/bucket.go:102 +0x15e
created by github.com/couchbaselabs/sync_gateway/base.couchbaseBucket.StartTapFeed
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/base/bucket.go:104 +0x1bd

goroutine 51 [IO wait]:
net.runtime_pollWait(0x7ffaca94dad0, 0x72, 0x0)
    /usr/local/go/src/pkg/runtime/netpoll.goc:116 +0x6a
net.(_pollDesc).Wait(0xc210151060, 0x72, 0x7ffaca94c0c0, 0xb)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:81 +0x34
net.(_pollDesc).WaitRead(0xc210151060, 0xb, 0x7ffaca94c0c0)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:86 +0x30
net.(_netFD).accept(0xc210151000, 0xab9170, 0x0, 0x7ffaca94c0c0, 0xb)
    /usr/local/go/src/pkg/net/fd_unix.go:382 +0x2c2
net.(_TCPListener).AcceptTCP(0xc2101beaa0, 0x18, 0xc2101ec050, 0x473f43)
    /usr/local/go/src/pkg/net/tcpsock_posix.go:233 +0x47
net.(_TCPListener).Accept(0xc2101beaa0, 0x50, 0x9b6800, 0x3, 0x9bf210)
    /usr/local/go/src/pkg/net/tcpsock_posix.go:243 +0x27
net/http.(_Server).Serve(0xc2101bc0f0, 0x7ffaca9573d8, 0xc2101beaa0, 0x0, 0x0)
    /usr/local/go/src/pkg/net/http/server.go:1622 +0x91
github.com/couchbaselabs/sync_gateway/base.ListenAndServeHTTP(0x9bf210, 0xe, 0x0, 0x0, 0x0, ...)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/base/http_listener.go:52 +0x3ee
github.com/couchbaselabs/sync_gateway/rest.(*ServerConfig).serve(0xc210083300, 0x9bf210, 0xe, 0x7ffaca94c2e0, 0xc2101e3cc0)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/rest/config.go:390 +0xa7
created by github.com/couchbaselabs/sync_gateway/rest.RunServer
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/src/github.com/couchbaselabs/sync_gateway/rest/config.go:428 +0x5ce

goroutine 52 [runnable]:
github.com/dustin/gomemcached/client.(_Client).runFeed(0xc2101a0600, 0xc210199ae0, 0xc2101d34c0)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/dustin/gomemcached/client/tap_feed.go:297 +0x5c3
created by github.com/dustin/gomemcached/client.(_Client).StartTapFeed
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/dustin/gomemcached/client/tap_feed.go:251 +0x215

goroutine 53 [select]:
github.com/couchbaselabs/go-couchbase.(_TapFeed).forwardTapEvents(0xc2101a1a00, 0xc2101d34c0, 0xc2102159c0, 0xc2101b4d80, 0x11)
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/tap.go:101 +0x249
created by github.com/couchbaselabs/go-couchbase.(_TapFeed).connectToNodes
    /data/jenkins/workspace/build_sync_gateway_103_centos-x64/app-under-test/sync_gateway/vendor/src/github.com/couchbaselabs/go-couchbase/tap.go:93 +0x367

**Comment by ajres** [comment]
@keshavkatwe 

Can you paste your Sync Gateway config to this ticket.

Andy

**Comment by keshavkatwe** [comment]
```
{
    "log": ["CRUD+", "REST+", "Changes+", "Attach+"],
    "databases": {
        "wishhapp": {
            "server": "http://127.0.0.1:8091",
            "bucket": "wishhapp",
            "shadow": {
               "server": "http://127.0.0.1:8091",
               "bucket": "default"
            },
            "sync": `
            function(doc){
                channel(doc.channels);
            }`,
            "users": {
                "GUEST": {"disabled": false, "admin_channels": ["*"]}
            }
        }
    }
}
```

**Comment by ajres** [comment]
@keshavkatwe 

Can you add "Shadow+" to the "log" property in the SG config and re-run.

The SG Log should contain a log entry just before the panic with the format:

```
Pulling %q, CAS=%x ... have UpstreamRev=%q, UpstreamCAS=%x"
```

This should show the document that is causing the panic.

If you can't see what the issue is from that, please post the log section with the entry above and the offending document from the shadow bucket (and the sg bucket if it already exists and the doc in the shadow bucket is a new revision)

Andy

**Comment by keshavkatwe** [comment]
Thank you, Got the issue. Some documents were null which caused the problem.

**Comment by ajres** [comment]
Update Sync Gateway to not panic on null documents in a shadow bucket

**Comment by adamcfraser** [comment]
@ajres Does the fix for #2102 also handle this case?

**Comment by adamcfraser** [comment]
Appears related to null document handling during bucket shadowing.  Closing as bucket shadowing is removed from product, and null docs are handled when using shared bucket access.
